vortex-scan/src/multi_scan.rs

```diff
@@ -46,44 +54,63 @@ impl<T> MultiScan<T> {
     ///
     /// The scan progresses when calling `next` on the iterator.
     pub fn new_scan_iterator(&self) -> MultiScanIterator<T> {
+        let worker = Worker::new_fifo();
+        self.stealers.write().push(worker.stealer());
+
         MultiScanIterator {
             scan_builder_factory: self.scan_builder_factory.clone(),
             local_pool: LocalPool::new(),
-            task_queue: SegQueue::new(),
+            stealers: self.stealers.clone(),
+            next_stealer_id: self.next_stealer_id.clone(),
+            worker,
         }
     }
 }
 
 /// Scan iterator to participate in a `MultiScan`.
 pub struct MultiScanIterator<T> {
     local_pool: LocalPool,
+    worker: Worker<ArrayFuture<T>>,
+    stealers: Arc<RwLock<Vec<Stealer<ArrayFuture<T>>>>>,
+    next_stealer_id: Arc<AtomicUsize>,
 
     /// Thread-safe queue of closures that lazily produce [`ScanBuilder`] instances.
     /// This queue is shared across all iterators being created with `new_scan_iterator`.
     scan_builder_factory: ScanBuilderFactory<T>,
-    task_queue: SegQueue<ArrayFuture<T>>,
 }
 
 impl<T: Send + Sync + 'static> Iterator for MultiScanIterator<T> {
     type Item = VortexResult<T>;
 
     fn next(&mut self) -> Option<VortexResult<T>> {
         // Queue up tasks if the thread local queue is empty.
-        if self.task_queue.is_empty() {
+        if self.worker.is_empty() {
             if let Some(scan_builder_fn) = self.scan_builder_factory.pop() {
                 match scan_builder_fn().build() {
                     Ok(tasks) => {
                         for task in tasks {
-                            self.task_queue.push(Box::pin(task));
+                            self.worker.push(Box::pin(task));
                         }
                     }
                     Err(err) => return Some(Err(err)),
                 }
+            } else {
+                let stealer_count = self.stealers.read().len();
+
+                for _ in 0..stealer_count {
```

Comment:
Something still feels super racy about this.

Multiple threads can increment the next_stealer_id, and each thread loops over only the number of stealers they witnessed prior to the loop. So it's definitely not guaranteed that each worker will try every stealer.

But will all stealers be tried by at least one worker?

And what if one worker pops the next scan builder, but hasn't pushed the tasks yet. So it has nothing to steal. The other workers see no more scans, and no tasks to steal, so they break out of the loop and terminate. Now we don't get work stealing? 

I'm kind of convinced this will always terminate... but not totally convinced it behaves the way we want.

Any ideas on making it more obviously correct? Otherwise we could try and fuzz it maybe.

Reply:
I was just thinking, maybe should always try to steal from the worker with the most items.

Reply:
I don't really see how correctness ties into this, rather than perf. As long as no work is completely dropped this is correct.

Reply:
> And what if one worker pops the next scan builder, but hasn't pushed the tasks yet. So it has nothing to steal.

Yes in that case they terminate too early. In the C++ version we tracked whether all tasks from all files had been created. We should do the same here.